src/signatures/schnorr.rs

```diff
@@ -70,6 +103,63 @@ where
         Ok(Self::new(public_nonce, s))
     }
 
+    /// Signs a message with the given secret key.
+    ///
+    /// This method correctly binds a nonce and the public key to the signature challenge, using domain-separated
+    /// hashing. The hasher is also opinionated in the sense that Blake2b 256-bit digest is always used.
+    ///
+    /// it is possible to customise the challenge by using [`construct_domain_separated_challenge`] and [`sign_raw`]
+    /// yourself, or even use [`sign_raw`] using a completely custom challenge.
+    pub fn sign_message<B>(secret: K, message: B) -> Result<Self, SchnorrSignatureError>
+    where
+        K: Add<Output = K> + Mul<P, Output = P> + Mul<Output = K>,
+        B: AsRef<[u8]>,
+    {
+        let nonce = K::random(&mut rand::thread_rng());
+        let public_nonce = P::from_secret_key(&nonce);
+        let public_key = P::from_secret_key(&secret);
+        let challenge = Self::construct_domain_separated_challenge::<_, Blake256>(&public_nonce, &public_key, message);
+        Self::sign_raw(secret, nonce, challenge.as_ref())
+    }
+
+    /// Constructs an opinionated challenge hash for the given public nonce, public key and message.
+    ///
+    /// In general, the signature challenge is given by `H(R, P, m)`. Often, plain concatenation is used to construct
+    /// the challenge. In this implementation, the challenge is constructed by means of domain separated hashing
+    /// using the provided digest.
+    ///
+    /// This challenge is used in the [`sign_message`] and [`verify_message`] methods.If you wish to use a custom
+    /// challenge, you can use [`sign_raw`] instead.
+    pub fn construct_domain_separated_challenge<B, D>(
+        public_nonce: &P,
+        public_key: &P,
+        message: B,
+    ) -> DomainSeparatedHash<D>
+    where
+        B: AsRef<[u8]>,
+        D: Digest,
+    {
+        DomainSeparatedHasher::<D, SchnorrSigChallenge>::new_with_label("challenge")
+            .chain(public_nonce.as_bytes())
+            .chain(public_key.as_bytes())
```

Comment:
You mean swapping the argument order? 

There can't be implementation confusion, since this is the implementation.

Reply:
I was thinking of the case where the developer uses the public `construct_domain_separated_challenge` as part of a custom use case. But on further thought that wouldn't really make sense, since you want to either use the safe wrapper functions (meaning you don't ever call this function) or handle the challenge yourself for a more complex use case (meaning you still wouldn't call this function).

Reply:
Yes, that's my thinking too.

FWIW, I was actually very torn about making `construct_domain_separated_challenge` public at all, but there might be times when you want to see how the sausage is made here, so I fell on the side of making it public,